File: backend/app/schemas/location.py
from pydantic import BaseModel, Field
from typing import Optional


class LocationUpdate(BaseModel):
    """Update user location"""
    city: Optional[str] = None
    latitude: Optional[float] = Field(None, ge=-90, le=90)
    longitude: Optional[float] = Field(None, ge=-180, le=180)


class CityGeocodeRequest(BaseModel):
    """Request to geocode a city name"""
    city: str = Field(..., min_length=2, max_length=255)
    country: Optional[str] = Field(None, max_length=100)


class CityGeocodeResponse(BaseModel):
    """Geocoded city result"""
    city: str
    country: str
    latitude: float
    longitude: float
    display_name: str

File: frontend/src/api/location.ts
import apiClient from './client';

export interface LocationUpdate {
  city?: string;
  latitude?: number;
  longitude?: number;
}

export interface CityGeocodeRequest {
  city: string;
  country?: string;
}

export interface CityGeocodeResponse {
  city: string;
  country: string;
  latitude: number;
  longitude: number;
  display_name: string;
}

export const locationApi = {
  // Update user location
  updateLocation: async (location: LocationUpdate): Promise<{message: string; city?: string; latitude?: number; longitude?: number}> => {
    const response = await apiClient.put('/users/location', location);
    return response.data;
  },

  // Geocode city name to coordinates
  geocodeCity: async (request: CityGeocodeRequest): Promise<CityGeocodeResponse> => {
    const response = await apiClient.post<CityGeocodeResponse>('/users/location/geocode', request);
    return response.data;
  },

  // Request browser geolocation
  requestBrowserLocation: (): Promise<GeolocationPosition> => {
    return new Promise((resolve, reject) => {
      if (!navigator.geolocation) {
        reject(new Error('Geolocation is not supported by your browser'));
        return;
      }

      navigator.geolocation.getCurrentPosition(
        (position) => resolve(position),
        (error) => reject(error),
        {
          enableHighAccuracy: true,
          timeout: 10000,
          maximumAge: 0
        }
      );
    });
  },
};
